client/go/dogma/watch_service.go

```diff
@@ -0,0 +1,358 @@
+
+	if latest := w.Latest(); latest.Err == nil {
+		go func() {
+			// Perform initial notification so that the listener always gets the initial value.
+			listener(latest.Revision, latest.Value)
+		}()
+	}
+
+	return nil
+}
+
+func (ws *watchService) fileWatcher(projectName, repoName string, query *Query) (*Watcher, error) {
+	if query == nil {
+		return nil, errors.New("query should not be nil")
+	}
+
+	w := newWatcher(projectName, repoName, query.Path)
+	w.doWatchFunc = func(lastKnownRevision int) <-chan *WatchResult {
+		return ws.watchFile(w.watchCTX, projectName, repoName, strconv.Itoa(lastKnownRevision),
+			query, watchTimeout)
+	}
+	w.convertingResultFunc = func(result *WatchResult) *Latest {
+		value := result.Commit.Entries[0].Content
+		return &Latest{Revision: result.Commit.Revision, Value: value}
+	}
+	return w, nil
+}
+
+func (ws *watchService) repoWatcher(projectName, repoName, pathPattern string) (*Watcher, error) {
+	w := newWatcher(projectName, repoName, pathPattern)
+	w.doWatchFunc = func(lastKnownRevision int) <-chan *WatchResult {
+		return ws.watchRepo(w.watchCTX, projectName, repoName, strconv.Itoa(lastKnownRevision),
+			pathPattern, watchTimeout)
+	}
+	w.convertingResultFunc = func(result *WatchResult) *Latest {
+		revision := result.Commit.Revision
+		return &Latest{Revision: revision, Value: revision}
+	}
+	return w, nil
+}
+
+func (w *Watcher) start() {
+	if atomic.CompareAndSwapInt32(&w.state, initial, started) {
+		w.scheduleWatch(0)
+	}
+}
+
+func (w *Watcher) scheduleWatch(numAttemptsSoFar int) {
+	if w.isStopped() {
+		return
+	}
+
+	var delay time.Duration
+	if numAttemptsSoFar == 0 {
+		latest := w.Latest()
+		if latest.Err != nil {
+			delay = delayOnSuccess
+		} else {
+			delay = 0
+		}
+	} else {
+		delay = nextDelay(numAttemptsSoFar)
+	}
+
+	go func() {
+		select {
+		case <-w.scheduleCTX.Done():
+		case <-time.NewTimer(delay).C:
+			w.doWatch(numAttemptsSoFar)
+		}
+	}()
+}
+
+func (w *Watcher) isStopped() bool {
+	state := atomic.LoadInt32(&w.state)
+	return state == stopped
+}
+
+func (w *Watcher) doWatch(numAttemptsSoFar int) {
+	if w.isStopped() {
+		return
+	}
+
+	var lastKnownRevision int
+	curLatest := w.Latest()
+	if curLatest.Revision == 0 {
+		lastKnownRevision = 1 // Init revision
+	} else {
+		lastKnownRevision = curLatest.Revision
+	}
+
+	select {
+	case <-w.watchCTX.Done():
+	case watchResult := <-w.doWatchFunc(lastKnownRevision):
+		if watchResult.Err != nil {
+			if watchResult.Err == context.Canceled {
+				// Cancelled by close()
+				return
+			}
+
+			log.Debug(watchResult.Err)
+			w.scheduleWatch(numAttemptsSoFar + 1)
+			return
+		}
+
+		newLatest := w.convertingResultFunc(watchResult)
+		if w.isInitialChanSet == 0 && atomic.CompareAndSwapInt32(&w.isInitialChanSet, 0, 1) {
+			// The initial latest is set for the first time. So write the value to initialChan as well.
+			w.initialChan <- newLatest
+		}
+		atomic.StorePointer(&w.latestPtr, (unsafe.Pointer)(newLatest))
```

Comment:
I'm not sure but could we use `atomic.Value.store/load` instead?

Reply:
I didn't know that I can use in that way. Thank you~!